File: Problem 2.py
from __future__ import print_function


def gen_fib(max_val):
    l1, l2 = 1,1
    nval = 0
    while nval < max_val:
        nval = sum([l1,l2])
        l1, l2 = l2, nval
        yield nval

if __name__ == "__main__":
    res = sum(i for i in gen_fib(4000000)
              if i & 1 == 0)
    print("The answer to Euler Problem 2 is {}".format(res))

File: Problem 2.hs
-- Slow naive fib
--fib 0 = 0
--fib 1 = 1
--fib n = fib (n-1) + fib(n-2)

-- fast memo'd fib from http://www.haskell.org/haskellwiki/Memoization
memoizedFib :: Int -> Integer
res :: Integer
main :: IO()

memoizedFib = (map fib [0 ..] !!)
   where fib 0 = 0
         fib 1 = 1
         fib n = memoizedFib (n-2) + memoizedFib (n-1)

res = sum  [y | y<-takeWhile (<4000000) [memoizedFib x | x <- [1..]], y `mod` 2 == 0]

main = print ("The answer to Euler Problem 2 is " ++ show res)